### PR TITLE
add `cf.grid_mapping_names`

### PR DESCRIPTION
**Pull request by larsbuntemeyer**
I wanted to pick up the discussion from https://github.com/xarray-contrib/cf-xarray/pull/370 of making `cf-xarray` understand the `grid_mapping` attribute. This alone (without the further interpretation by `pyproj`) would be really useful to RCM datasets (you never know if the mapping is called `rotated_pole`, `rotated_latitude_longitude` or whatever...). 

The implementation is straight forward, the [logic in the CF conventions](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.10/cf-conventions.html#grid-mappings-and-projections) mainly follows the bounds logic. A data variable can have a `grid_mapping` attribute that points to another data variable containing the `grid_mapping_name` attribute. As far as i know, the name of the variable holding the `grid_mapping_name` attribute might be arbitrary. Actually, it's a lot of code copied from the `bounds` logic, which i should optimize, sorry... Just wanted to throw this in quickly, let me know if it's useful...

- [x] add `cf.grid_mappings`
- [x] update repl
- [x] add a new page to the docs titled "Grid Mappings"
- [x] add tests
- [x]  add DataArray tests 
- [ ] ~~update _COORD_NAMES and _AXIS_NAMES~~
- [x] Add more See Also sections
- [x] Update docs

Here's an example:
```python
from cf_xarray.datasets import rotds
rotds.cf
```
```
Coordinates:
- CF Axes: * X: ['rlon']
           * Y: ['rlat']
             Z, T: n/a

- CF Coordinates:   longitude: ['lon']
                    latitude: ['lat']
                    vertical, time: n/a

- Cell Measures:   area, volume: n/a

- Standard Names: * grid_latitude: ['rlat']
                  * grid_longitude: ['rlon']
                    latitude: ['lat']
                    longitude: ['lon']

- Bounds:   n/a

- Grid Mappings:   n/a

Data Variables:
- Cell Measures:   area, volume: n/a

- Standard Names:   air_temperature: ['temp']

- Bounds:   lat: ['lat_bounds']
            latitude: ['lat_bounds']
            lon: ['lon_bounds']
            longitude: ['lon_bounds']

- Grid Mappings:   air_temperature: ['rotated_pole']
                   temp: ['rotated_pole']
```
and...
```
rotds.cf.grid_mappings
```
```
{'air_temperature': ['rotated_pole'], 'temp': ['rotated_pole']}
```
access it:
```python
rotds.cf.get_grid_mapping("temp")
```
grafik
get name of the mapping (might be different to the data variable name):
```python
rotds.cf.get_grid_mapping_name("temp")
```
```
'rotated_latitude_longitude'
```

### CODE REVIEW

File: cf_xarray/accessor.py [hunk 1]
```diff
@@ -2335,6 +2363,82 @@ def bounds_to_vertices(
...
+
+        results = self.grid_mappings.get(key, [])
+        if not results:
+            raise KeyError(f"No results found for {key!r}.")
+
+        return self._obj[results[0] if len(results) == 1 else results]
+
+    def get_grid_mapping_name(self, key: str) -> str:
+        """
+        Get bounds dim name for variable corresponding to key.
```
Comment (dcherian) [inline cf_xarray/accessor.py:2422]: Let's update this docstring.


In general, this one seems redundant (`ds.cf.get_grid_mapping('temp").attrs["grid_mapping_name"]` seems not too bad) but happy to defer to an actual user :)

**Review by dcherian — COMMENTED** [review]
Thanks for contributing @larsbuntemeyer 

This is looking good to me. I have a few requests:
1. Please add a new page to the docs titled "Grid Mappings"
2. Please add to `api.rst`
3. Please add tests
4. Should we add a specific `DataArray` property that's more narrow? It seems like it could work since Xarray propagates scalars with all DataArrays when extracted.
5. That reminds me, we should update `get_associated_variables` too and check that `grid_mapping` information is propagated with `ds.cf["temp"]`

(3), (4) could happen in a separate PR if you prefer.

The tests failues can be fixed by updating the "expected" repr

**Comment by dcherian** [comment]
@jthielen do you have any thoughts on this API, and/or suggestions for API additions that would be useful.

**Comment by larsbuntemeyer** [comment]
> 1. Please add a new page to the docs titled "Grid Mappings"
> 
> 2. Please add to `api.rst`
> 
> 3. Please add tests

Great, thanks! I'll start polishing!

> 5. That reminds me, we should update `get_associated_variables` too and check that `grid_mapping` information is propagated with `ds.cf["temp"]`

Yes, that would be great. I saw that in `datasets.airds` you explicitly added `cell_area` to `coords`. That should also happen for the `grid_mapping` i guess.

> (3), (4) could happen in a separate PR if you prefer.

I'll add tests here...

**Comment by dcherian** [comment]
Seems like we should add `grid_latitude` and `grid_longitude` to `_COORD_NAMES` too? This would allow `ds.cf.coords["grid_latitude"]` for example

**Comment by larsbuntemeyer** [comment]
Yes, that would be nice! `grid_latitude` and `grid_longitude` are somehow specific to `rotated_latitude_longitude`. There is also:

* `projection_x_angular_coordinate` and `projection_y_angular_coordinate` for geostationary projection
* `projection_x_coordinate` and `projection_y_coordinate` for all other mappings

Should i add them in this PR?

**Comment by dcherian** [comment]
> Should i add them in this PR?

Sure. I think @jthielen may have commented earlier about not adding these, but I can't find it now.

**Comment by larsbuntemeyer** [comment]
maybe #23 is related.

**Comment by dcherian** [comment]
> maybe https://github.com/xarray-contrib/cf-xarray/issues/23 is related.

Thanks that's right. I think adding to `_COORD_NAMES` should be fine. `projection_x,y,angular_coordinate` should perhaps be under `_AXIS_NAMES`.

This is quite minor though,

File: cf_xarray/accessor.py (hunk 1 above)
Comment (larsbuntemeyer) [reply cf_xarray/accessor.py]: That's true! I just thought, having this function explictily in the API would raise some awareness for the difference between `grid_mapping` and `grid_mapping_name` attributes (especially if they are different)!

**Comment by larsbuntemeyer** [comment]
I am still not quite sure how to handle propagation...

File: cf_xarray/accessor.py [hunk 2]
```diff
@@ -2335,6 +2372,88 @@ def bounds_to_vertices(
                 )
         return obj
 
+    @property
+    def grid_mappings(self) -> dict[str, list[str]]:
+        """
+        Property that returns a dictionary mapping keys
```
Comment (dcherian) [inline cf_xarray/accessor.py:2378]:
```suggestion
        Property that returns a dictionary mapping variable names
```

File: cf_xarray/accessor.py [hunk 3]
```diff
@@ -2335,6 +2372,88 @@ def bounds_to_vertices(
...
+        to the variable names of their grid_mapping.
+
+        Returns
+        -------
+        dict
+            Dictionary mapping keys to the variable names of their grid_mapping.
```
Comment (dcherian) [inline cf_xarray/accessor.py:2384]:
```suggestion
            Dictionary mapping variable names to the variable names of their grid_mapping.
```

File: cf_xarray/accessor.py [hunk 4]
```diff
@@ -1593,10 +1622,15 @@ def get_associated_variable_names(
...
-        keys = ["ancillary_variables", "cell_measures", "coordinates", "bounds"]
+        keys = [
+            "ancillary_variables",
+            "cell_measures",
+            "coordinates",
+            "bounds",
+            "grid_mapping",
+        ]
         coords: dict[str, list[str]] = {k: [] for k in keys}
         attrs_or_encoding = ChainMap(self._obj[name].attrs, self._obj[name].encoding)
```
Comment (dcherian) [inline cf_xarray/accessor.py:1633]:
```suggestion
        attrs_or_encoding = ChainMap(self._obj[name].attrs, self._obj[name].encoding)

```

It's nice to minimize any unnecessary changes.

File: cf_xarray/datasets.py [hunk 5]
```diff
@@ -225,6 +225,10 @@ def _create_mollw_ds():
 def _create_inexact_bounds():
     # Dataset that creates rotated pole curvilinear coordinates with CF bounds in
     # counterclockwise order that have precision issues.
+    # dataset created using: https://gist.github.com/larsbuntemeyer/105d83c1eb39b1462150d3fabca0b66b
```
Comment (dcherian) [inline cf_xarray/datasets.py:228]: Very nice. In a future PR, it would be nice to add this as a "demo" notebook which creates a datasets, and illustrates creation of bounds/ extracting grid_mapping etc.

File: doc/grid_mappings.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+```
+
+## Use `grid_mapping` in projections
+
+The grid mapping information use very useful in projections, e.g., for plotting. [pyproj](https://pyproj4.github.io/pyproj/stable/api/crs/crs.html#pyproj.crs.CRS.from_cf) understands CF conventions right away, e.g.
+
+```python
+from pyproj import CRS
+
+CRS.from_cf(rotds.cf.get_grid_mapping("temp").attrs)
```
Comment (dcherian) [inline doc/grid_mappings.md:49]: Hmm... it seems like it'd be nice to have some kind of `GridMapping` class that had `as_pyproj` and `as_cartopy`. If you're up for it, I encourage you to pursue such a discussion somewhere upstream.

Alternatively, if we could also have `DataArray.cf.get_pyproj_crs()` and work to get cartopy understand that.

Can you open an issue to discuss this? We don't need to resolve it now.

 cc @aulemahal

File: doc/grid_mappings.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+- native coordinates in which the data is defined, e.g., regular 1D coordinates
+- projected coordinates which probably denote some "real" coordinates in [latitude and longitude](https://en.wikipedia.org/wiki/Geographic_coordinate_system#Latitude_and_longitude)
+
+Due to the projection, those real coordinates are probably 2D data variables. The `grid_mapping` attribute of a data variable makes a connection to another data variable defining the coordinate reference system (CRS) of those native coordinates. It should enable you to project the native coordinates into any other CRS, including the real 2D latitude and longitude coordinates. This is often useful for plotting, e.g., you can [tell cartopy how to correctly plot coastlines](https://scitools.org.uk/cartopy/docs/latest/tutorials/understanding_transform.html) for the CRS your data is defined in.
+
+To access `grid_mapping` attributes, consider this example:
+
+```{code-cell}
+from cf_xarray.datasets import rotds
```
Comment (dcherian) [inline doc/grid_mappings.md:32]:
```suggestion
from cf_xarray.datasets import rotds

```

**Review by dcherian — COMMENTED** [review]
Let's not add the projection coordinates to the repr yet. We can deal with it when someone has a use case :)

For propagating info when extracting a DataArray, can you add a failing test please? I can take a look then.

File: doc/grid_mappings.md (hunk 6 above)
Comment (aulemahal) [reply doc/grid_mappings.md:49]: That would be nice indeed!

AFAIU, this is sadly still partially broken : https://github.com/SciTools/cartopy/issues/2099. The CRS returned by this function doesn't work with `cartopy`, which would be the main use case, I believe.

Comment (larsbuntemeyer) [reply doc/grid_mappings.md:49]: Yes, absolutely, that would be a nice feature. I guess, that's what @aulemahal is heading for with https://github.com/SciTools/cartopy/issues/2099.

File: cf_xarray/accessor.py [hunk 8]
```diff
@@ -2335,6 +2372,88 @@ def bounds_to_vertices(
...
+        to the variable names of their grid_mapping.
+
+        Returns
+        -------
+        dict
+            Dictionary mapping variable names to the variable names of their grid_mapping.
+
+        See Also
+        --------
+        Dataset.cf.get_grid_mapping_name
```
Comment (dcherian) [inline cf_xarray/accessor.py:2388]:
```suggestion
        Dataset.cf.get_grid_mappings
        Dataset.cf.get_grid_mapping_name
```

File: cf_xarray/accessor.py [hunk 9]
```diff
@@ -2335,6 +2372,88 @@ def bounds_to_vertices(
...
+    def get_grid_mapping_name(self, key: str) -> str:
+        """
+        Get name of the grid_mapping for variable corresponding to key.
+
+        Parameters
+        ----------
+        key : str
+            Name of variable whose grid_mapping name is desired.
+
+        Returns
```
Comment (dcherian) [inline cf_xarray/accessor.py:2444]: Please add a `See Also` section

File: cf_xarray/accessor.py [hunk 10]
```diff
@@ -2335,6 +2372,88 @@ def bounds_to_vertices(
...
+        """
+        Get grid_mapping variable corresponding to key.
+
+        Parameters
+        ----------
+        key : str
+            Name of variable whose grid_mapping is desired
+
+        Returns
+        -------
```
Comment (dcherian) [inline cf_xarray/accessor.py:2425]: Please add a "See Also" section

File: cf_xarray/tests/test_accessor.py [hunk 11]
```diff
@@ -935,6 +956,57 @@ def test_get_bounds_dim_name():
...
+    """
+    assert dedent(expected) in ds.cf.__repr__()
+
+    # Extraction
+    assert ds.cf["temp"].coords["rotated_pole"] == rotds.rotated_pole
+
+    # DataArray
+    # propagation does not work yet properly
+    actual = ds.cf["temp"].cf.__repr__()
+    assert actual == expected
```
Comment (larsbuntemeyer) [inline cf_xarray/tests/test_accessor.py:986]: > For propagating info when extracting a DataArray, can you add a failing test please? I can take a look then.

I think this should pass, but i am not sure why it's not working properly. The rotated_pole is extracted into coords but no seen in the repr...

Comment (dcherian) [reply cf_xarray/tests/test_accessor.py:986]: I think you need to add a specific `grid_mappings` (maybe singular for DataArray) attribute so `ds["temp"].cf.grid_mapping` works.

File: cf_xarray/accessor.py [hunk 12]
```diff
@@ -1379,6 +1406,7 @@ def make_text_section(subtitle, attr, valid_values=None, default_keys=None):
             )
             text += make_text_section("Standard Names", "standard_names", data_vars)
             text += make_text_section("Bounds", "bounds", data_vars)
+            text += make_text_section("Grid Mappings", "grid_mappings", data_vars)
```
Comment (dcherian) [inline cf_xarray/accessor.py:1409]: Does this make sense for Data Variables too?

File: cf_xarray/accessor.py [hunk 13]
```diff
@@ -1635,6 +1675,10 @@ def get_associated_variable_names(
                 if dbounds:
                     coords["bounds"].append(dbounds)
 
+        if not skip_grid_mappings:
+            if "grid_mapping" in attrs_or_encoding:
+                coords["grid_mapping"] = [attrs_or_encoding["grid_mapping"]]
+
```
Comment (dcherian) [inline cf_xarray/accessor.py:1681]: I don't know that we need `skip_grid_mappings` unless you have a use case.

`bounds` are funny because sometimes bounds have the same `standard_name` and in some cases, we know the user doesn't want that.

File: doc/grid_mappings.md (hunk 6 above)
Comment (dcherian) [reply doc/grid_mappings.md:50]: Is there a workaround we can use?

File: cf_xarray/accessor.py [hunk 14]
```diff
@@ -2335,6 +2376,101 @@ def bounds_to_vertices(
...
+        ----------
+        Please refer to the CF conventions document : https://cfconventions.org/Data/cf-conventions/cf-conventions-1.10/cf-conventions.html#grid-mappings-and-projections
+
+        For a list of valid grid_mapping names, refer to: https://cfconventions.org/Data/cf-conventions/cf-conventions-1.10/cf-conventions.html#appendix-grid-mappings
+
+        Examples
+        --------
+        >>> from cf_xarray.datasets import rotds
+        >>> rotds.cf.grid_mappings
+        {'air_temperature': ['rotated_pole'], 'temp': ['rotated_pole']}
```
Comment (dcherian) [inline cf_xarray/accessor.py:2405]:
```suggestion
	{'rotated_pole": ['air_temperature', 'temp']}
```

So, our current properties (axes, standard_names, bounds, etc.)have this format : "CF name": "variable name". 

I think we should do that instead. To get the values for a single DataArray, the user can still do `ds.cf["air_temperature"].cf.grid_mapping`

File: cf_xarray/accessor.py [hunk 15]
```diff
@@ -2510,6 +2646,20 @@ def formula_terms(self) -> dict[str, str]:
             terms[key] = value
         return terms
 
+    @property
+    def grid_mapping(self) -> DataArray | None:
```
Comment (dcherian) [inline cf_xarray/accessor.py:2650]: Could also add a `grid_mapping_name` property to get the string if that seems useful to  you.

Comment (larsbuntemeyer) [reply cf_xarray/accessor.py:2650]: Yes, i think, that makes sense here!

File: cf_xarray/accessor.py (hunk 14 above)
Comment (larsbuntemeyer) [reply cf_xarray/accessor.py:2405]: Agreed! I didnt' really recognized that logic from the beginning...

Comment (larsbuntemeyer) [reply cf_xarray/accessor.py:2439]: Should it maybe even be `{'rotated_latitude_longitude": ['air_temperature', 'temp']}`, that's actually the valid `grid_mapping_name` for rotated pole.

File: cf_xarray/tests/test_accessor.py (hunk 11 above)
Comment (larsbuntemeyer) [reply cf_xarray/tests/test_accessor.py:1001]: > For bonus points (future PR is OK) I think we should handle the repr for grid_mapping specifically by surfacing extra info like the actual mapping name, and potential other useful info.

I think that would require some adaptations to `make_text_section`, right? I am a little undecided what to do with the repr for grid mappings as a coordinate. `ds["temp"].cf.grid_mapping` now returns a DataArray, no dictionary...

File: cf_xarray/accessor.py (hunk 14 above)
Comment (dcherian) [reply cf_xarray/accessor.py:2439]: Ah yes that seems right

File: cf_xarray/tests/test_accessor.py (hunk 11 above)
Comment (dcherian) [reply cf_xarray/tests/test_accessor.py:1001]: #391 makes some big repr changes, so we can just skip this for now.

File: cf_xarray/accessor.py (hunk 14 above)
Comment (larsbuntemeyer) [reply cf_xarray/accessor.py:2439]: ok, after all, when if understand correctly, it should work like

* `rotds.cf.grid_mappings` should give: `{'rotated_latitude_longitude": ['rotated_pole']}`
* `rotds.cf["rotated_latitude_longitude"]` should return `rotated_pole` DataArray.

The connection between data variables and grid mappings would work through the DataArray accessors.
Sorry, i was just following the bounds logic in the beginning (which seemed similar)....

Comment (dcherian) [reply cf_xarray/accessor.py:2440]: Is it common to have more than one `grid_mapping` for a Dataset? 

We could also make `DataArray.cf["grid_mapping"] -> DataArray` work and `Dataset.cf["grid_mapping"] -> Dataset` work. 

But we can leave it for later too.

Comment (larsbuntemeyer) [reply cf_xarray/accessor.py:2440]: It's rather uncommon to have more than one grid_mapping actually! I just felt that it should be possible to access by `grid_mapping_name` (which also follows a little bit the measures logic). I actually like `{'rotated_latitude_longitude": ['air_temperature', 'temp']}` because it gives a good overview!

Comment (dcherian) [reply cf_xarray/accessor.py:2440]: > it should be possible to access by grid_mapping_name

Yup I agree.

It requires you to know the actual grid mapping, which is why I suggested `.cf['grid_mapping']` . But yeah, we can wait for someone to request it.

File: cf_xarray/tests/test_accessor.py (hunk 11 above)
Comment (dcherian) [reply cf_xarray/tests/test_accessor.py]: Sorry i meant #409

**Comment by larsbuntemeyer** [comment]
Ok, this seems to work now, i'll summarize:

* `rotds.cf.grid_mappings` returns `{'rotated_latitude_longitude': ['rotated_pole']}`
*  `rotds.cf["rotated_latitude_longitude"]` returns `rotds.rotated_pole`
*  `rotds.cf["temp"].cf.grid_mapping` returns `rotds.rotated_pole`
*  `rotds.cf["temp"].cf.grid_mapping_name` returns `rotated_latitude_longitude`

I would remove `Dataset.cf.get_grid_mapping` and `Dataset.cf.get_grid_mapping_name` in favour of the dataarray accessors and a future `Dataset.cf["grid_mapping"]` and update docs....

**Comment by dcherian** [comment]
I changed it so that:

* `rotds.cf.grid_mapping_names` returns `{'rotated_latitude_longitude': ['rotated_pole']}`
*  `rotds.cf["rotated_latitude_longitude"]` returns `rotds.rotated_pole`
*  `rotds.cf["temp"].cf.grid_mapping_name` returns `'rotated_latitude_longitude'`

I enabled the `.cf["grid_mapping"]` which will return a DataArray, so that removes the need for `DataArray.cf.grid_mapping`. Sorry about this, I need to write it down somewhere, but in general, properties return strings and `__getitem__` returns arrays. So here I set it so that `"grid_mapping"` will get us the DataArray.


In the future, we might think about adding a custom `GridMapping` class, returned by `DataArray.cf.grid_mapping` that has methods to convert it to WKT or pyproj.CRS or whatever.

Please try it out with your code and see if you like it.

**Comment by larsbuntemeyer** [comment]
Thanks a lot, that's really awesome and makes sense to me, will be really useful for working with CORDEX data! I did some minor updates for the docs!

**Comment by larsbuntemeyer** [comment]
I guess, the error is caused by the CF tables being updated just yesterday!

**Comment by dcherian** [comment]
Thanks @larsbuntemeyer this is great!

> will be really useful for working with CORDEX data! 

If you ever write a demo notebook showcasing this, it would be good to link to it in our docs. That would be a valuable community contribution!

**Comment by larsbuntemeyer** [comment]
> If you ever write a demo notebook showcasing this, it would be good to link to it in our docs.

Absolutely! Thanks for the support!
